Product Query - add support for the global query

gigitux opened the pull request:
This PR is blocked by woocommerce/woocommerce-blocks#7257.

This PR adds:
- an option equivalent to “Inherit Query from Template” (perhaps even visually identical) to make the Product Query compatible with FSE templates. ⚠️ Please, don't review the UI and the label. We can iterate them easily in the next weeks.
- Merge global query into variation query.
- Merge query parameters from the URL into the global query to allow compatibility with filters.

Fixes woocommerce/woocommerce-blocks#6863

<!-- Don't forget to update the title with something descriptive. -->
<!-- If you can, add the appropriate labels -->


#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](https://github.com/woocommerce/woocommerce-blocks/blob/trunk/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

<!-- If your change has a visual component, add a screenshot here. A "before" screenshot would also be helpful. -->

![image](https://user-images.githubusercontent.com/4463174/195354190-f54742de-c0b1-4007-9e43-c4e10a4fef77.png)


### Testing

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

<!-- Write these steps from the perspective of a "user" (merchant) familiar with WooCommerce. No need to spell out the steps for common setup scenarios (eg. "Create a product"), but do be specific about the thing being tested. Include screenshots demonstrating expectations where that will be helpful. -->

### Product Search Template

1. Edit the `Product Search` template.
2. In the header section, add the `Product Search` block.
3. Replace the PHP classic block with the Product Query block.
4. Set the `woo inherit query from template` setting to true.
5. Save. Go on the front end.
6. With the `Product Search` block, search a product and be sure that the results that the Product Query renders are correct.
7. Edit the `Product Search` template.
8. Add filter blocks or/and enable the `on-sale` setting. Go on the front end.
9. With the `Product Search` block, search a product and add some filters and be sure that the results that the Product Query renders are correct.

### Products By Category Template

1. Edit the `Products by Category` template.
2. Replace the PHP classic block with the Product Query block.
3. Set the `woo inherit query from template` setting to true.
4. Save. Go on the front end.
5. Go to this URL `product-category/clothing/tshirts/`. In this way, the `Product by Category` will be rendered and should be rendered only products that are in the T-shirts category.
7. Edit the `Products By Category` template.
8. Add filter blocks or/and enable the `on-sale` setting. Go on the front end.
9. Go to this URL `product-category/clothing/tshirts/`. In this way, the `Product by Category` will be rendered and should be rendered only products that are in the T-shirts category and match the filters/variation.

### Products By Tag Template

1. Edit the `Products by Tag` template.
2. Replace the PHP classic block with the Product Query block.
3. Set the `woo inherit query from template` setting to true.
4. Create a tag and add some products.
5. Save. Go on the front end.
6. Go to this URL `/product-tag/${name_tag}/`. In this way, the `Product by Tag` will be rendered and should be rendered only products that are in the selected tag.
7. Edit the `Products By Tag` template.
8. Add filter blocks or/and enable the `on-sale` setting. Go on the front end.
9. Go to this URL `/product-tag/${name_tag}/`. In this way, the ` Products By Tag` template will be rendered and should be rendered only products that are in the selected tag and match the filters/variation.



* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see if the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [ ] WooCommerce Core
* [ ] Feature plugin
* [x] Experimental

gigitux committed:
Product Query: Fix pagination issue

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/product-query-update

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/price_filter_support

gigitux committed:
Product Query - Add support for the Filter By Price Block #6790

Product Query - Add support for the Filter By Price Block

gigitux committed:
fix query relation

gigitux committed:
fix on sale query

gigitux committed:
Merge branch 'trunk' into add/price_filter_support

gigitux committed:
Merge branch 'trunk' into add/price_filter_support

gigitux committed:
Product Query - Add support for the Filter By Attributes block #6790

Product Query - Add support for the Filter By Attributes block

gigitux committed:
Merge branch 'trunk' into add/attribute_filter_support

gigitux committed:
fix bugged pagination and on-sale filter after refactor

gigitux committed:
Merge branch 'add/price_filter_support' of https://github.com/woocommerce/woocommerce-blocks into add/price_filter_support

gigitux committed:
Merge branch 'add/price_filter_support' of https://github.com/woocommerce/woocommerce-blocks into add/attribute_filter_support

gigitux committed:
Merge branch 'add/attribute_filter_support' of https://github.com/woocommerce/woocommerce-blocks into add/attribute_filter_support

gigitux committed:
Merge ranch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/attribute_filter_support

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/attribute_filter_support

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into add/attribute_filter_support

gigitux committed:
address feedback

gigitux committed:
Product Query - Add support for the Filter By Stock Block #6790

Product Query - Add support for the Filter By Stock Block

gigitux committed:
Fix filter blocks: the data (e.g: max price or stock-status) match the variation #7245

fix filter blocks: the data (e.g: max price or stock-status) match the variation

gigitux committed:
Merge branch 'trunk' into fix/filter_display_data

gigitux committed:
Merge branch 'trunk' into fix/filter_display_data

gigitux committed:
Merge branch 'trunk' into fix/filter_display_data

gigitux committed:
Merge branch 'trunk' into fix/filter_display_data

gigitux committed:
rename allowControls to allowedControls

gigitux committed:
Merge branch 'fix/allowed-controls-product-query' of https://github.com/woocommerce/woocommerce-blocks into fix/product-query-global-query

gigitux committed:
add support to the global query

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/product-query-global-query

gigitux committed:
Merge branch 'trunk' into fix/product-query-global-query

gigitux committed:
fix eslint error

gigitux committed:
Merge branch 'fix/product-query-global-query' of https://github.com/woocommerce/woocommerce-blocks into fix/product-query-global-query

gigitux committed:
Merge branch 'trunk' into fix/product-query-global-query

gigitux committed:
Merge branch 'trunk' into fix/product-query-global-query

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/product-query-global-query

gigitux committed:
Merge branch 'fix/product-query-global-query' of https://github.com/woocommerce/woocommerce-blocks into fix/product-query-global-query

dinhtungdu commented on src/BlockTypes/ProductQuery.php:
```diff
@@ -455,4 +470,27 @@ function( $acc, $array ) {
 	}
 
+	/**
+	 * Return a query that filters products by the global query.
```
```suggestion
	 * Get product-related query variables from the global query.
```

dinhtungdu commented on src/BlockTypes/ProductQuery.php:
```diff
@@ -455,4 +470,27 @@ function( $acc, $array ) {
+		}
+
+		return array(
+			'taxonomy' => isset( $wp_query->query_vars['taxonomy'] ) ? $wp_query->query_vars['taxonomy'] : '',
+			'term'     => isset( $wp_query->query_vars['term'] ) ? $wp_query->query_vars['term'] : '',
```
`taxonomy` and `term` need to be parsed into a `tax_query` array item. See how Core does: https://github.com/WordPress/wordpress-develop/blob/7ecb5239ef8d2133e63c003810e20f2e971a70c8/src/wp-includes/class-wp-query.php#L1146-L1152

dinhtungdu replied:
Checking `Woo Inherit query from template`, I got an empty query on the front end of product category pages.

gigitux replied:
> taxonomy and term need to be parsed into a tax_query array item. 

Good catch! I updated it! In any case, with the previous version, I was able to see the right query 🤔

dinhtungdu commented on src/BlockTypes/ProductQuery.php:
```diff
@@ -455,4 +470,27 @@ function( $acc, $array ) {
+	 * @return array
+	 */
+	private function get_global_query( $parsed_block ) {
+		global $wp_query;
```
I think we should also check if we're on a WooCommerce page before processing. For Product Query, we care about the Product Catalog (`is_shop()`) and product taxonomy archive pages (`is_product_taxonomy()`).

gigitux replied:
Mmm, not sure about this check. We want to use the Product Query for the Product Search template too.

In any case, I would say not to add any particular check here. In this way, 3rd developers could use the Product Query everywhere. The important thing is that they update the global `wp_query`.

dinhtungdu commented on src/BlockTypes/ProductQuery.php:
```diff
@@ -133,31 +139,54 @@ public function build_query( $query ) {
 				return $this->merge_queries( $acc, $query );
 			},
-			$common_query_values
+			$base_query
```
Here we merge the global query into queries by attributes and filters, which means that for example:

A Product Query block that inherits the query from the `Hoodies` category page can also display the products from the `Tshirts` category.

I don't think this is the correct behavior. We should honor the global query, we can add new query but shouldn't modify the query provided by the global query:
- On the product category page, we shouldn't modify the `tax_query`.
- But we can add filters, for example, showing only In Stock products from the current category.

gigitux replied:
Mmmm, not sure about this! I think that when the global query is enabled, we should merge some attributes.


> A Product Query block that inherits the query from the Hoodies category page can also display the products from the Tshirts category.

For example, it makes sense that we don't want to allow this, but at the same time, we want that the user can use `Inherit query from template` with the `on-sale` attribute, right?

dinhtungdu replied:
> but at the same time, we want that the user can use `Inherit query from template` with the `on-sale` attribute, right?

Yes, users can choose to display only on-sale products or only in-stock products on the Hoodies category page, but all products must be in the Hoodies category. The same applies to tag and product attribute archives.

gigitux replied:
It already works this way. Did you find a bug? 


https://user-images.githubusercontent.com/4463174/200870619-93394b09-2835-4d9a-baf0-15d09442f10f.mp4

There are some issues related to the inherent query that I will address in dedicated PRs #7609 #7608

dinhtungdu reviewed (commented):
@gigitux Thanks for working on this PR. I have some comments after testing and reviewing the diff, please see more details in the review comments.

gigitux committed:
Merge branch 'trunk' into fix/product-query-global-query

github-actions[bot] committed:
bot: update checkstyle.xml

gigitux committed:
Update src/BlockTypes/ProductQuery.php

Co-authored-by: Tung Du <dinhtungdu@gmail.com>

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/product-query-global-query

gigitux committed:
fix global query generation

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/product-query-global-query

gigitux committed:
Merge branch 'trunk' into fix/product-query-global-query

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/product-query-global-query

dinhtungdu reviewed (commented):
@gigitux I'm sorry for this late review!

Followed the testing instructions, I can confirm this PR is testing well. Adding Product Query block to product archive templates with inherit query from template enabled, the block displays the products, but with some issues:
- The order of products is different between Classic template and Product Query. We're inheriting only search queries and tax queries, we need to inherit order and orderby variables too.

<details>
<summary>Screenshot of different results between Product Query and Classic Template</summary>

<img width="1912" alt="image" src="https://user-images.githubusercontent.com/5423135/201043948-96c21b37-d7b9-4c51-b6ca-ee3853049cd1.png">

</details>

- The `posts_per_page` and `paged` variables belong to the global query. Continue using the example above, adding both Product Query and Classic Template blocks side by side to the Product Category template:
    - Classic template display 12 products (WooCommerce settings).
    - Product Query display 6 products (Block setting).
    - Product Query still handles the next pages at the block level
    	- Expected: `http://wc.local/product-category/clothing/page/2/`
    	- Current: `http://wc.local/product-category/clothing/?query-0-page=2`

- After enabling inheriting, users can still add taxonomy filters, which makes the query return no results. We shouldn't allow users to add product category filters inside the Products by Category template if `Inherit query from template` is enabled.

<details>
<summary>Screenshot of adding product category filter to Product Query inside Product by Category template</summary>

<img width="1248" alt="image" src="https://user-images.githubusercontent.com/5423135/201040144-a155b3a9-478f-450a-b6b3-9d7265d4d28d.png">

</details>

- These issues above raise another question about extensibility, themes and extensions can add custom query variables, how should we cover that?

gigitux commented:
Thanks for your review and your precious feedback, @dinhtungdu! The goal of this PR is to be the foundation of this feature. As you said, there are other issues that we should take care of, but I think that we should address them in dedicated PRs, with specific E2E tests, to avoid any future regression.

> The order of products is different between Classic template and Product Query. We're inheriting only search queries and tax queries, we need to inherit order and orderby variables too.

If we follow this pattern, I think that we should disable these settings: 
![image](https://user-images.githubusercontent.com/4463174/201054504-c9e2e779-2d4d-43be-99ff-71bc0ee0f987.png)

Same concept for the number of products. There is a setting, should we disable it? I think we should wait for the final UI/UX mockup and, after that, take this kind of decision.

Also, I created a draft PR to fix the default order of the products https://github.com/woocommerce/woocommerce-blocks/pull/7641

There is a dedicated issue for this issue https://github.com/woocommerce/woocommerce-blocks/issues/7626

> Product Query still handles the next pages at the block level
Expected: http://wc.local/product-category/clothing/page/2/
Current: http://wc.local/product-category/clothing/?query-0-page=2

This is a good question. To be honest, I didn't think about this issue. Do you think that it is something that we should address? What happens if multiple Product Query blocks are on the same page? cc @sunyatasattva 

> After enabling inheriting, users can still add taxonomy filters, which makes the query return no results. We shouldn't allow users to add product category filters inside the Products by Category template if Inherit query from template is enabled.

UI/UX issue. We can easily address it in the following PRs.


> These issues above raise another question about extensibility, themes and extensions can add custom query variables, how should we cover that?

For the first iteration, the extensibility isn't a goal. I think we should investigate it in the next phase of the project.

dinhtungdu commented:
> Do you think that it is something that we should address?

Given that product archive templates will be powered by Product Query, this is something we have to address.

Correct me if I'm wrong, but I think we may need to change the direction of inheriting the global query.
- We are **getting some** global query variables and merging them with our customer product filters and attributes.
- But I think we should: **start with** global query and merge a selective set of our custom product filters and attributes, in other world, take the whole global query as the starting point and work from there.

gigitux commented:
> * But I think we should: **start with** global query and merge a selective set of our custom product filters and attributes, in other world, take the whole global query as the starting point and work from there.

Mmm, good point! I agree with you! I'm not sure that WP_Query can parse the global query (I think that I never tried), so I thought that was easier and sure the first direction. We can try your suggestion!

sunyatasattva commented:
Just chiming in here.

I don't have a strong opinion on how to implement this specifically, but just to make sure we are on the same page regarding the requirements:

The hierarchy of the query should be as such: Global Query → Merchant block settings → User-facing filters. It is basically like a matter of CSS specificity: user filters should always have the last word on what's displayed, of course; and merchants should be able to override anything in the Global Query.

Perhaps we should make it so that the defaults are in line between the Classic Templates and the Product Query (e.g. show 12 items per page by default instead of 6), but I think that there is no reason to limit merchants when it comes to deciding how many items per page they want to show.

I think we agree on this, but I am not sure, so I'm asking.


Regarding the URL issue:

> Product Query still handles the next pages at the block level
> Expected: http://wc.local/product-category/clothing/page/2/
> Current: http://wc.local/product-category/clothing/?query-0-page=2

Well spotted @dinhtungdu! This is indeed a super tough question. I think that, given that we want multiple Query blocks on one page, we can't support the first style of URL, as we can't decide which one is the “main” one. Unless we want to navigate ALL of them. I think this is going to be less relevant when hydration comes, as it is less confusing for the user.

For SEO, as long as we SSR, it should be fine.

dinhtungdu commented:
> I think that, given that we want multiple Query blocks on one page, we can't support the first style of URL, as we can't decide which one is the “main” one. Unless we want to navigate ALL of them.

@sunyatasattva The main one is the one inheriting query from the template. If we have more than one Product Query blocks that inherit the query from the template, then we have multiple main ones, and we should navigate ALL of them.

> but I think that there is no reason to limit merchants when it comes to deciding how many items per page they want to show.

I 100% agree, we shouldn't limit merchants to do that. I'll be so sad if merchants still need to go to the setting page or use PHP filters to change the number of products per page for 'blockified archive templates'.
I'm thinking about these potential directions:
- Add posts per page setting for archive templates.
- Sync posts per page setting of the Product Query block back to the template query. All Product Query blocks that inherit global query in the same template will share the same posts per page setting, changing the setting in one block will affect others. (I don't feel safe with this approach, instead of passing data down to the nested objects, we allow nested objects to affect the ancestors.

gigitux committed:
Merge branch 'trunk' into fix/product-query-global-query

github-actions[bot] committed:
bot: update checkstyle.xml

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/product-query-global-query

gigitux committed:
Merge branch 'fix/product-query-global-query' of https://github.com/woocommerce/woocommerce-blocks into fix/product-query-global-query

dinhtungdu committed:
try: refactor merge_queries

dinhtungdu committed:
remove debug code

dinhtungdu committed:
Merge branch 'trunk' into refactor/merge_queries

dinhtungdu committed:
Merge branch 'trunk' into refactor/merge_queries

dinhtungdu committed:
Merge branch 'trunk' into refactor/merge_queries

dinhtungdu committed:
fix: unpack assoc array php 7.4

dinhtungdu committed:
try: refactor merge_queries to take any form of input without unpacking
and preparing the input arrays

dinhtungdu committed:
add missing query vars

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/product-query-global-query

gigitux committed:
Merge branch 'refactor/merge_queries' of https://github.com/woocommerce/woocommerce-blocks into fix/product-query-global-query

gigitux committed:
Merge branch 'trunk' into fix/product-query-global-query

gigitux committed:
add feature flag

gigitux commented:
Hey @dinhtungdu,
I updated this PR and I added the feature flag (on both sides, maybe should we do it only on the JS side?)! Also, in this PR I already merged woocommerce/woocommerce-blocks#7697.

dinhtungdu commented on assets/js/blocks/product-query/variations/product-query.tsx:
```diff
@@ -19,6 +19,11 @@ import {
 
 const VARIATION_NAME = 'woocommerce/product-query';
 
+// This is a feature flag to enable the custom inherit Global Query implementation.
+// This is not intended to be a permanent feature flag, but rather a temporary.
+// https://github.com/woocommerce/woocommerce-blocks/pull/7382
+const isCustomInheritGlobalQueryImplementationEnabled = false;
```
What do you think about using an experimental [flag](https://github.com/woocommerce/woocommerce-blocks/blob/2646666a8ebe1e4813e98087268048dfcd9c3445/docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md?plain=1#L21-L23) here instead? I think we just need to limit the global query support to experimental/development build only. 🤔

gigitux replied:
> I think we just need to limit the global query support to experimental/development build only.

At least for now, we want to show the option inherited query from the template from Gutenberg. The Product Query is a foundation for other blocks, too (for example, `Buy it again`), and having `two inherit query from template` options can be confusing for the developers that check the block for the first time.

Also, this is a super WIP code because there is still a discussion about how we should approach the problem: this is why I used this approach.

For me, It is fine to change it, but I think it could not be very clear.

dinhtungdu replied:
> The Product Query is a foundation for other blocks, too (for example, Buy it again), and having two inherit query from template options can be confusing for the developers that check the block for the first time.

Make senses!

Can you explain how can we use this flag for development later? I mean what do we need to do to enable the flag? At glance, we have to temporarily change the flag in the source code, right?

gigitux replied:
> Can you explain how can we use this flag for development later? I mean what do we need to do to enable the flag? At glance, we have to temporarily change the flag in the source code, right?

Yes! Just set `isCustomInheritGlobalQueryImplementationEnabled` and `is_custom_inherit_global_query_implementation_enabled` to true.

dinhtungdu replied:
Maybe adding a note here to update the PHP flag too? (and vice-versa)

gigitux replied:
It makes sense! I updated the comments!

gigitux committed:
Merge branch 'trunk' into fix/product-query-global-query

dinhtungdu reviewed (approved):
Thanks for the explanation! I left an optional comment but as discussed, this is LGTM so approving!

gigitux committed:
Merge branch 'trunk' of https://github.com/woocommerce/woocommerce-blocks into fix/product-query-global-query

gigitux committed:
updates comment

gigitux committed:
Merge branch 'fix/product-query-global-query' of https://github.com/woocommerce/woocommerce-blocks into fix/product-query-global-query
